Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## 1. Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('/datasets/insurance.csv')

In [3]:
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  5000 non-null int64
Возраст              5000 non-null float64
Зарплата             5000 non-null float64
Члены семьи          5000 non-null int64
Страховые выплаты    5000 non-null int64
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [5]:
target = df['Страховые выплаты']

In [6]:
features = df.drop('Страховые выплаты', axis=1)

### Получили на входе таблицу с данными клиентов страхавой компании. Имеем пять колонок из них признаки: пол, возраст, зарплата застрахованного, количество членов его семьи. Целевой признак - количество страховых выплат клиенту за последние 5 лет. Всего предоставленны данные по 5000 клиентов.

## 2. Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

Опишем процесс умножения матриц.

Чтобы матрицу   можно было умножить на матрицу  нужно, чтобы число столбцов матрицы  равнялось числу строк матрицы. В нашем случае необходимо умножить признаки на обратимую матрицу размера 4х4. 

Ответ: При умножении признаков на обратимую матрицу качество линейной регрессии не изменится.

Обоснование: Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

При умножении признаков на матрицу получим следующее выражение
$$
w'=((XP)^T(XP))^{−1}(XP)^Ty
$$
вынесем Р
$$
w'=(P^TX^TXP)^{−1}X^TP^Ty
$$


$$
w'=(P^TP)^{−1}(X^TX)^{−1}X^TP^Ty
$$


$$
w'=P^{−1}(P^T)^{−1}(X^TX)^{−1}X^TP^Ty
$$
в итоге получим, используя свойство $(P^T)^{−1}P^T=E$

Е - еденичная матрица
$$
w'=(X^TX)^{−1}P^{−1}X^Ty
$$
вектор весов линейной регрессии преобразованой матрицы будет иметь значение 𝑤′=𝑤(𝑃)−1 и значение предсказания примет вид
𝑎=X^PwP^{−1}
после сокращения 𝑃^{−1}𝑃=Е имем прежний вид значения предсказания и следовательно умножение признаков на матрицу не окажет влияния на качество линейной регрессии.

## 3. Алгоритм преобразования

**Алгоритм**

Обозначения:

- $X$ — значения признаков 
- $P$ — матрица, на которую умножаются признаки

Для преобразования данных необходимо умножить признаки на обратимую матрицу. Обратимая матрица возможна только для прямоугольных матриц. Получим матрицу квадратного размера использовав функцию rand которая может быть использована для генерации двумерных массивов. Затем обязательно проверим полученную матрицу на обратимость использовав np.linalg.inv(). Если в процессе выполнения функции не возникнет ошибки, то это будет подтверждать обратимость полученной матрицы. При отрицательном результате необходимо повторно создать матрицу. 

Далее умножив исходные данные с иходными признаками на обратимую матрицу получим преобразованые признаки обеспечивающие защиту данных клиентов 
$$
P*X
$$

**Обоснование**

Данный метод пробразования можно применять вследствии того что качество линейной регрессии основаной на исходных данных и преобразованных данных не изменится. 

## 4. Проверка алгоритма

In [7]:
# обучим модель линейной регресии и проверим метрику R2 на необработанных данных
model = LinearRegression()
model.fit(features, target)
predictions = model.predict(features)
print(r2_score(target, predictions))

0.42494550286668


In [8]:
X = features

In [9]:
# создадим  матрицу 
P = np.random.rand(4,4)

In [10]:
# проверим матрицу на обратимость, если не будет ошибки то матрица обратима
try:
    P_1 = np.linalg.inv(P)
except:
    print('Матрица не прошла проверку на обратимость, повторите шаг создания матрицы')

In [11]:
# умножин признаки на обратимую матрицу
X_modern = X @ P

In [12]:
# обучим модель линейной регресии и проверим метрику R2 на обработанных данных
model = LinearRegression()
model.fit(X_modern, target)
predictions = model.predict(X_modern)
print(r2_score(target, predictions))

0.4249455028666206


### Вывод. Обучив модель линейной регрессии на исходных и преобразованных данных признаков видим что метрика R2 в обоих случаях составила около 0.4249 - т.е. качество модели не изменилось после преобразования и выбор метода обработки персональных данных клиентов был выбран правильно.